include/dgl/immutable_graph.h

```diff
@@ -699,95 +820,78 @@ class ImmutableGraph: public GraphInterface {
    * \param fmt the format of the returned adjacency matrix.
    * \return a vector of three IdArray.
    */
-  virtual std::vector<IdArray> GetAdj(bool transpose, const std::string &fmt) const;
+  std::vector<IdArray> GetAdj(bool transpose, const std::string &fmt) const override;
 
-  /*
-   * The immutable graph may only contain one of the CSRs (e.g., the sampled subgraphs).
-   * When we get in csr or out csr, we try to get the one cached in the structure.
-   * If not, we transpose the other one to get the one we need.
-   */
-  CSR::Ptr GetInCSR() const {
-    if (in_csr_) {
-      return in_csr_;
-    } else {
-      CHECK(out_csr_ != nullptr) << "one of the CSRs must exist";
-      const_cast<ImmutableGraph *>(this)->in_csr_ = out_csr_->Transpose();
-      return in_csr_;
+  /* !\brief Return in csr. If not exist, transpose the other one.*/
+  CSRPtr GetInCSR() const {
+    if (!in_csr_) {
+      if (out_csr_) {
+        in_csr_ = out_csr_->Transpose();
```

Comment:
is transpose from out csr more efficient than conversion from coo directly?

Reply:
I think they are similar in terms of complexity. But for cache locality, I'm not sure.

Reply:
i see how csr transpose and coo=>csr works now. in the worst case, both should have the same cache locality. However, coo=>csr might work better in practice because coo is already sorted. i'm not sure either if this is a common case.

Reply:
coo is not sorted in src order but in eid order, so one could expect the locality is pretty bad